### PR TITLE
LAU-574 Create POST /tasks/delete Endpoint

### PR DESCRIPTION
**Pull request by MartinYSpasov**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/LAU-574

### Change description ###
Create POST /tasks/delete Endpoint

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by rafalkalita — CHANGES_REQUESTED** [review]
Looks good. Just few lines not covered by tests. Please check.

**Review by rafalkalita — COMMENTED** [review]
As per comment in the https://tools.hmcts.net/jira/browse/LAU-574 Ravi suggested this should be put behind feature toggle

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/TaskActionsController.java [hunk 1]
```diff
@@ -341,6 +350,43 @@ public ResponseEntity<GetTaskRolePermissionsResponse> getTaskRolePermissions(
...
+            if (!hasAccess) {
+                return buildErrorResponseEntityAndLogError(HttpStatus.FORBIDDEN.value(),
+                        new GenericForbiddenException(GENERIC_FORBIDDEN_ERROR));
+            }
+
+            verifyCaseId(deleteTasksRequest.getDeleteCaseTasksAction().getCaseRef());
+
+            taskDeletionService.deleteTasksByCaseId(deleteTasksRequest.getDeleteCaseTasksAction().getCaseRef());
+
+            return status(HttpStatus.CREATED.value())
```
Comment (kvvasudev) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/TaskActionsController.java:377]: I think the status code has to be OK than CREATED ? As the business logic underneath is to remove data from the db.

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/request/DeleteCaseTasksAction.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import lombok.ToString;
+
+import java.io.Serializable;
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@EqualsAndHashCode
+@ToString
+public class DeleteCaseTasksAction implements Serializable {
+
+    public static final long serialVersionUID = 432973322;
```
Comment (kvvasudev) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/request/DeleteCaseTasksAction.java:17]: Can an Unique SerialVersionID be created please ?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/request/DeleteTasksRequest.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+@Schema(
+        name = "DeleteTasksRequest",
+        description = "Allows specifying certain completion options"
+)
+@EqualsAndHashCode
+@ToString
+public class DeleteTasksRequest implements Serializable {
+
+    public static final long serialVersionUID = 432973322;
```
Comment (kvvasudev) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/request/DeleteTasksRequest.java:20]: Can an Unique SerialVersionID be created please ?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepository.java [hunk 4]
```diff
@@ -62,6 +63,10 @@ public interface TaskResourceRepository extends CrudRepository<TaskResource, Str
 
     List<TaskResource> findAllByTaskIdIn(List<String> taskIds, Sort order);
 
+    @Query(value = "select c.task_id AS taskid, c.state AS state from {h-schema}tasks c where c.case_id=:caseId",
```
Comment (kvvasudev) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepository.java:66]: Just to follow the conventions, can this query be moved to a constant ?

File: src/test/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/TaskActionsControllerTest.java [hunk 5]
```diff
@@ -427,6 +435,46 @@ void should_return_task_roles_with_permissions() {
         assertEquals(response.getBody().getPermissionsList(), taskRolePermissions);
     }
 
+
+    @Test
+    void should_return_201_response_for_tasks_deletion() {
```
Comment (kvvasudev) [inline src/test/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/TaskActionsControllerTest.java:440]: May be need to change the signature to 200 than 201 ?

File: src/test/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskDatabaseServiceTest.java [hunk 6]
```diff
@@ -395,4 +395,16 @@ void should_find_task_to_update_index_return_list_of_tasks() {
         assertNotNull(actualTaskResource);
         assertEquals(1, actualTaskResource.size());
     }
+
+
+    @Test
+    void should_delete_tasks_by_task_ids() {
+        final List<String> taskIds = List.of("123", "456");
+        lenient().doNothing().when(taskResourceRepository).deleteAllById(taskIds);
```
Comment (kvvasudev) [inline src/test/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskDatabaseServiceTest.java:403]: Do we need lenient here ?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/TaskActionsController.java (hunk 1 above)
Comment (MartinYSpasov) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/TaskActionsController.java:377]: The status code is defined in our LLD and agreed with Nigel Barlass - https://tools.hmcts.net/confluence/pages/diffpages.action?originalId=1660290997&pageId=1663966928&__ncforminfo=R7E3CoIvliTg-67RCXoKZ2UoRheYVnEIclYSS1XGCZSRzGnJxzxeY7SO3GWVbFBcVS7U5Tlaus9OTgv_gjBsCWxoMBcw1Q6p3faUIWHmMas=

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/request/DeleteTasksRequest.java (hunk 3 above)
Comment (MartinYSpasov) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/request/DeleteTasksRequest.java:20]: Done

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/request/DeleteCaseTasksAction.java (hunk 2 above)
Comment (MartinYSpasov) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/request/DeleteCaseTasksAction.java:17]: Done

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepository.java (hunk 4 above)
Comment (MartinYSpasov) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepository.java:66]: Done

File: src/test/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskDatabaseServiceTest.java (hunk 6 above)
Comment (MartinYSpasov) [reply src/test/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskDatabaseServiceTest.java:403]: Done

File: src/test/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/TaskActionsControllerTest.java (hunk 5 above)
Comment (MartinYSpasov) [reply src/test/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/TaskActionsControllerTest.java:440]: Please refer to my previous comment

**Review by SammiChong — DISMISSED** [review]
LGTM

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskDeletionService.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    @Autowired
+    public TaskDeletionService(final CFTTaskDatabaseService cftTaskDatabaseService) {
+        this.cftTaskDatabaseService = cftTaskDatabaseService;
+    }
+
+    public void deleteTasksByCaseId(final String caseId) {
+        final List<TaskResourceCaseQueryBuilder> taskResourceCaseQueryBuilders = cftTaskDatabaseService
+                .findByTaskIdsByCaseId(caseId);
+
+        filterAllUnterminatedTasksAndLogError(taskResourceCaseQueryBuilders, caseId);
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskDeletionService.java:28]: As discussed, could this be updated to reflect that the "delete" logging in  filterAllUnterminatedTasksAndLogError happens after we call  deleteTasks(taskResourceCaseQueryBuilders, caseId);

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskDeletionService.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+                                                               taskResourceCaseQueryBuilders,
+                                                       final String caseId) {
+        final List<String> unterminatedTaskIds = taskResourceCaseQueryBuilders
+                .stream()
+                .filter(task -> !task.getState().equals(TERMINATED))
+                .map(TaskResourceCaseQueryBuilder::getTaskId)
+                .toList();
+
+        if (!unterminatedTaskIds.isEmpty()) {
+            log.error(String.format("Deleted some UNTERMINATED tasks: %s for caseId: %s", unterminatedTaskIds, caseId));
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskDeletionService.java:53]: As this is a valid edge case for tracking unterminated tasks, we can use the OutputCaptured to update our new delete integration tests to prove our logging captures and logs this for this kind of scenario. 

There is an example of using CapturedOutput on this integration test -> ExecuteReconfigurationTaskOperationControllerTest.should_execute_reconfigure_on_task_and_fail_due_to_calendar_configuration_alerts_captured

**Review by BenNzewi — COMMENTED** [review]
We are missing the pact provider and consumer tests for the new endpoint. Please raise this within the retain and dispose QA. We will need the following implemented as high priority under the contractTest package;

1. delete Provider on the WA side
2. delete local consumer on the WA side (see existing action endpoint example)
3.  Consumer on the Retain and Dispose service

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/utils/InputParamsVerifier.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+import static org.apache.commons.lang3.StringUtils.isEmpty;
+import static uk.gov.hmcts.reform.wataskmanagementapi.services.utils.DeleteTaskConstants.CASEID_DELETE_EXCEPTION_MESSAGE;
+
+public final class InputParamsVerifier {
+
+    private InputParamsVerifier() {
+    }
+
+    public static void verifyCaseId(final String caseRef) {
+        if (isEmpty(caseRef)
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/utils/InputParamsVerifier.java:17]: As discussed, for a new line we need unit test coverage.  See ref -> https://sonarcloud.io/component_measures?metric=new_coverage&selected=uk.gov.hmcts.reform%3Awa-task-management-api%3Asrc%2Fmain%2Fjava%2Fuk%2Fgov%2Fhmcts%2Freform%2Fwataskmanagementapi%2Fcontrollers%2Futils%2FInputParamsVerifier.java&view=list&pullRequest=991&id=uk.gov.hmcts.reform%3Awa-task-management-api

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskDeletionService.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+        deleteTasks(taskResourceCaseQueryBuilders, caseId);
+    }
+
+    private void deleteTasks(final List<TaskResourceCaseQueryBuilder> taskResourceCaseQueryBuilders,
+                             final String caseId) {
+        try {
+            cftTaskDatabaseService.deleteTasks(getTaskIds(taskResourceCaseQueryBuilders));
+        } catch (final Exception exception) {
+            log.error(String.format("Unable to delete all tasks for case id: %s", caseId));
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskDeletionService.java:38]: Please see missing coverage for new lines as discussed - > https://sonarcloud.io/component_measures?metric=new_coverage&selected=uk.gov.hmcts.reform%3Awa-task-management-api%3Asrc%2Fmain%2Fjava%2Fuk%2Fgov%2Fhmcts%2Freform%2Fwataskmanagementapi%2Fservices%2FTaskDeletionService.java&view=list&pullRequest=991&id=uk.gov.hmcts.reform%3Awa-task-management-api

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskDeletionService.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    private void filterAllUnterminatedTasksAndLogError(final List<TaskResourceCaseQueryBuilder>
+                                                               taskResourceCaseQueryBuilders,
+                                                       final String caseId) {
+        final List<String> unterminatedTaskIds = taskResourceCaseQueryBuilders
+                .stream()
+                .filter(task -> !task.getState().equals(TERMINATED))
+                .map(TaskResourceCaseQueryBuilder::getTaskId)
+                .toList();
+
+        if (!unterminatedTaskIds.isEmpty()) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskDeletionService.java:52]: Please see missing coverage for new lines as discussed -> https://sonarcloud.io/component_measures?metric=new_coverage&selected=uk.gov.hmcts.reform%3Awa-task-management-api%3Asrc%2Fmain%2Fjava%2Fuk%2Fgov%2Fhmcts%2Freform%2Fwataskmanagementapi%2Fservices%2FTaskDeletionService.java&view=list&pullRequest=991&id=uk.gov.hmcts.reform%3Awa-task-management-api

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/DeleteTasksByCaseIdControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+                ENDPOINT_BEING_TESTED,
+                deleteTasksRequest,
+                caseworkerCredentials.getHeaders()
+        );
+
+        result.then().assertThat().statusCode(HttpStatus.CREATED.value());
+    }
+
+    @Test
+    public void should_return_400_when_when_case_id_incorrect() {
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/DeleteTasksByCaseIdControllerTest.java:52]: BadRequest is covered in the new integration tests as well. We can remove this scenario from the FT

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/DeleteTasksControllerTest.java [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+        when(authTokenGenerator.generate())
+                .thenReturn(IDAM_AUTHORIZATION_TOKEN);
+        mockServices = new ServiceMocks(
+                idamWebApi,
+                serviceAuthorisationApi,
+                camundaServiceApi,
+                roleAssignmentServiceApi
+        );
+    }
+
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/DeleteTasksControllerTest.java:69]: As discussed, we should use any of the candidate tests here for the CapturedOutput where we are verifying our logging of untermintated tasks being deleted.

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskDeletionService.java (hunk 7 above)
Comment (MartinYSpasov) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskDeletionService.java:28]: done

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskDeletionService.java (hunk 8 above)
Comment (MartinYSpasov) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskDeletionService.java:53]: Done

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/DeleteTasksByCaseIdControllerTest.java (hunk 12 above)
Comment (MartinYSpasov) [reply src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/DeleteTasksByCaseIdControllerTest.java:52]: Done

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/DeleteTasksControllerTest.java (hunk 13 above)
Comment (MartinYSpasov) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/DeleteTasksControllerTest.java:69]: Done

Comment (MartinYSpasov) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/DeleteTasksControllerTest.java:75]: Done

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/DeleteTasksByCaseIdControllerTest.java (hunk 12 above)
Comment (MartinYSpasov) [reply src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/DeleteTasksByCaseIdControllerTest.java]: Done

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskDeletionService.java (hunk 11 above)
Comment (MartinYSpasov) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskDeletionService.java:51]: Done

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskDeletionService.java (hunk 10 above)
Comment (MartinYSpasov) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskDeletionService.java:37]: Done

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/utils/InputParamsVerifier.java (hunk 9 above)
Comment (MartinYSpasov) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/utils/InputParamsVerifier.java:17]: Done

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskDeletionService.java (hunk 8 above)
Comment (MartinYSpasov) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskDeletionService.java:52]: Done

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskDeletionService.java (hunk 7 above)
Comment (MartinYSpasov) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskDeletionService.java]: Done

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/TaskActionsController.java [hunk 14]
```diff
@@ -341,6 +350,43 @@ public ResponseEntity<GetTaskRolePermissionsResponse> getTaskRolePermissions(
...
+                    .cacheControl(CacheControl.noCache())
+                    .build();
+
+        } catch (final InvalidRequestException invalidRequestException) {
+            return buildErrorResponseEntityAndLogError(HttpStatus.BAD_REQUEST.value(), invalidRequestException);
+        } catch (final GenericForbiddenException invalidRequestException) {
+            return buildErrorResponseEntityAndLogError(HttpStatus.FORBIDDEN.value(), invalidRequestException);
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/TaskActionsController.java:384]: We can remove this one catch for Forbidden. It is handled above. line 369

**Review by BenNzewi — APPROVED** [review]
